mlx_graphs/datasets/planetoid.py

```diff
@@ -0,0 +1,201 @@
+) -> Data:
+    items = [read_file(folder, prefix, name) for name in file_names]
+    node_features, tx, allx, y, ty, ally, graph, test_index = items
+    train_index = mx.arange(y.shape[0], dtype=mx.int32)
+    val_index = mx.arange(y.shape[0], y.shape[0] + 500, dtype=mx.int32)
+    sorted_test_index = mx.sort(test_index)
+
+    if prefix.lower() == "citeseer":
+        # There are some isolated nodes in the Citeseer graph, resulting in
+        # none consecutive test indices. We need to identify them and add them
+        # as zero vectors to `tx` and `ty`.
+        len_test_indices = (test_index.max() - test_index.min()).item() + 1
+
+        tx_ext = mx.zeros((len_test_indices, tx.shape[1]), dtype=tx.dtype)
+        tx_ext[sorted_test_index - test_index.min(), :] = tx
+        ty_ext = mx.zeros((len_test_indices, ty.shape[1]), dtype=ty.dtype)
+        ty_ext[sorted_test_index - test_index.min(), :] = ty
+
+        tx, ty = tx_ext, ty_ext
+
+    else:
+        node_features = mx.concatenate([allx, tx], axis=0)
+        node_features[test_index] = node_features[sorted_test_index]
+
+    y = mx.argmax(mx.concatenate([ally, ty], axis=0), axis=1)
+    y[test_index] = y[sorted_test_index]
+
+    train_mask = index_to_mask(train_index, size=y.shape[0])
+    val_mask = index_to_mask(val_index, size=y.shape[0])
+    test_mask = index_to_mask(test_index, size=y.shape[0])
+
+    edge_index = edge_index_from_dict(
+        graph_dict=graph,
+        num_nodes=y.shape[0],
+        without_self_loops=without_self_loops,
+    )
+
+    graph = GraphData(
+        edge_index,
+        node_features,
+        node_labels=y.astype(mx.int32),
+    )
+    graph.train_mask = train_mask
+    graph.val_mask = val_mask
+    graph.test_mask = test_mask
+
+    return graph
+
+
+def read_file(folder: str, prefix: str, name: str) -> mx.array:
+    path = osp.join(folder, name)
+    prefix = prefix.lower()
+
+    if name == f"ind.{prefix}.test.index":
+        return read_txt_array(path, dtype=mx.int32)
+
+    with fsspec.open(path, "rb") as f:
+        warnings.filterwarnings("ignore", ".*`scipy.sparse.csr` name.*")
+        out = pickle.load(f, encoding="latin1")
+
+    if name == f"ind.{prefix}.graph":
+        return out
+
+    out = out.todense() if hasattr(out, "todense") else out
+    out = mx.array(out.tolist(), dtype=mx.float32)
+    return out
+
+
+def edge_index_from_dict(
+    graph_dict: dict[int, list[int]],
+    num_nodes: Optional[int] = None,
+    without_self_loops: bool = True,
+) -> mx.array:
+    rows: List[int] = []
+    cols: List[int] = []
+    for key, value in graph_dict.items():
+        rows += repeat(key, len(value))
+        cols += value
+
+    row = mx.array(rows)
+    col = mx.array(cols)
+    edge_index = mx.stack([row, col], axis=0)
+
+    # NOTE: There are some duplicated edges and self loops in the datasets.
+    #       Other implementations do not remove them!
```

Comment:
are the features of the duplicated edges identical?

Reply:
I guess, as it's done like this in PyG.